src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -0,0 +1,731 @@
+        while (is_directory(dir_ / tmpDir));
+
+        auto d = dir_ / tmpDir;
+        auto config {config_};
+        config.set("path", d.string());
+        {
+            fdLimit_ = f->createInstance(NodeObject::keyBytes,
+                config, scheduler_, j_)->fdlimit();
+            backed_ = static_cast<bool>(fdLimit_);
+        }
+        remove_all(d);
+    }
+    if (!backed_)
+        return true;
+
+    // Find shards
+    for (auto const& d : directory_iterator(dir_))
+    {
+        if (!is_directory(d))
+            continue;
+        auto dirName = d.path().stem().string();
+        if (!std::all_of(dirName.begin(), dirName.end(), ::isdigit))
+            continue;
+        auto const shardIndex {std::stoul(dirName)};
+        if (shardIndex < detail::genesisShardIndex)
+            continue;
+        auto shard = std::make_unique<Shard>(
+            shardIndex, cacheSz_, cacheAge_, j_);
+        if (!shard->open(config_, scheduler_, dir_))
+            return false;
+        usedDiskSpace_ += shard->fileSize();
+        if (shard->complete())
+            complete_.emplace(shard->index(), std::move(shard));
+        else
+        {
+            if (incomplete_)
+            {
+                JLOG(j_.error()) <<
+                    "More than one control file found";
+                return false;
+            }
+            incomplete_ = std::move(shard);
+        }
+    }
+    if (!incomplete_ && complete_.empty())
+    {
+        // New Shard Store, calculate file descriptor requirements
+        if (maxDiskSpace_ > space(dir_).free)
+        {
+            JLOG(j_.warn()) << "Insufficient disk space";
+        }
+        fdLimit_ = 1 + (fdLimit_ *
+            std::max<std::uint64_t>(1, maxDiskSpace_ / avgShardSz_));
+    }
+    else
+        updateStats();
+    return true;
+}
+
+boost::optional<std::uint32_t>
+DatabaseShardImp::prepare(std::uint32_t validLedgerSeq)
+{
+    std::lock_guard<std::mutex> l(m_);
+    if (incomplete_)
+        return incomplete_->prepare();
+    if (!canAdd_)
+        return boost::none;
+    if (backed_)
+    {
+        // Create a new shard to acquire
+        if (usedDiskSpace_ + avgShardSz_ > maxDiskSpace_)
+        {
+            JLOG(j_.debug()) <<
+                "Maximum size reached";
+            canAdd_ = false;
+            return boost::none;
+        }
+        if (avgShardSz_ > boost::filesystem::space(dir_).free)
+        {
+            JLOG(j_.warn()) <<
+                "Insufficient disk space";
+            canAdd_ = false;
+            return boost::none;
+        }
+    }
+
+    auto const shardIndexToAdd = findShardIndexToAdd(validLedgerSeq);
+    if (!shardIndexToAdd)
+    {
+        JLOG(j_.debug()) <<
+            "No new shards to add";
+        canAdd_ = false;
+        return boost::none;
+    }
+    // With every new shard, clear family caches
+    app_.shardFamily()->reset();
+    int const sz {std::max(shardCacheSz, cacheSz_ / std::max(
```

Comment:
What is the reasoning behind  `cacheSz_/ std::max(...`?

Reply:
Each shard owns a `TaggedCache` and its size is set by the `node_size` config setting. The memory requirement grows in relation to the number of shards and can be quite taxing with the larger settings. This simply tries to spread the desired setting across shard caches. Ideally, we need a unified cache shared by all but that is a project on its own. I've added that to my roadmap.